packages/mobx-state-tree/src/core/node/object-node.ts

```diff
@@ -495,17 +494,16 @@ export class ObjectNode<C, S, T> extends BaseNode<C, S, T> {
 
     @action
     die(): void {
-        if (this.state === NodeLifeCycle.DETACHING) return
-        if (this._observableInstanceState === ObservableInstanceLifecycle.CREATED) {
-            this.aboutToDie()
-            this.finalizeDeath()
-        } else {
-            // get rid of own and child ids at least
-            this.unregisterIdentifiers()
```

Comment:
this hack is not needed anymore since now array children are properly destroyed when poped/shifted